Add Sample Documentation (#55)
LM-Development/aks-sample: Samples/PublicSamples/RecordingBot/docs/explanations/recording-bot-overview.md

# High Level Overview

![Image 1,](../images/Overview.svg)

> As seen in the image Teams clients communicate with the Teams Server / Teams Communication Platform. When a user under compliance recording policy initiates a call (a call with another user or a call into a meeting) in its teams client, the Teams platform gets the HTTPS notification URL from a bot service that is linked to the Compliance Recording Policy. The Teams Platform uses the notification URL to do a authorized http call to the Recording Application, informing about the user initiating a call, authorization is done with a JWT token in the headers.

Reviewer: Clarity and capitalization. Avoid long sentences. Could be written as a sequence.

> [!IMPORTANT]
> The notification URL must be HTTPS and the certificate used must be signed by a valid authority.

> When the Recording Application receives the HTTPS Request, that a user is about to join a call, from the Teams Platform, the Recording Application should send an answer to the Teams Platform via the Graph Communications API within this answer the Recording Application has to give a TCP endpoint, can specify configuration and can change the notification url for further notifications regarding this call (the call the user initiated, to other users or into a meeting) that should be recorded. Further Notifications are send for everytime a user joins, mutes, starts a screen sharing and a lot of more actions users can do in a meeting.

Reviewer: Shorter sentences. send != sent. everytime != every time. Capitalization

> The TCP endpoint also requires a certificate signed by a valid authority. The Certificate for the HTTPS endpoint and the TCP endpoint **can** be the same.

## Graph Communications SDK

The overhead of the TCP endpoints is completly managed by the Graph Communications SDK. But the HTTPS endpoints for notifications must be custom implemented by the bot recording application, ASP.NET Core can be used. After a HTTP request is authorized and notifications are parsed from the request, the notifications should be forwarded to the SDK. It can then process them and trigger event handlers based on the notifications. The event handlers can implement business logic. For Example an event handler, that triggers when a new call notification was received, should be used for answering calls. Before answering business logic can decide whether the call should be accepted. Such an event handler should use the `answer` method of the SDK with the desired configuration to accept a call. The configuration specifies how many video sockets should be created, if the notification url should change and more. In the case that a call shouldn't be accepted a `reject` method of the SDK can be used.

> As the Microsoft Graph Communications SDK takes care of a lot of overhead and endpoint handling, the SDK also needs to be configured correctly: with a valid certificate as part of the TCP endpoint configuration with DNS name and port(s), the configuration of the notification Endpoints and configuration for the connection to the Graph Communications API. The SDK also needs a implentation of an authorization handler for outgoing API calls to the Graph Communications API, and for validating incoming requests, within the initial TCP handshake of the SDK and authorizing HTTP requests on the notification endpoint SDK. For the latter it's not required to use the handler but a simple possibility.

Reviewer 2: ```suggestion
As the Microsoft Graph Communications SDK takes care of a lot of overhead and endpoint handling, the SDK also needs to be configured with: 
- a valid certificate as part of the TCP endpoint configuration
- a DNS name
- port(s),
- notification Endpoints configuration
- Graph Communications API connection configuration.
The SDK also needs an implementation of an authorization handler for authoring
- requests against the Graph Communications API
- incoming requests (within the initial TCP handshake of the SDK)
- (optionally) HTTP requests on the notification endpoint of the SDK.
```

Teams platform, the application should answer the call via the Graph Communications API. The API
request to answer must contain some configuration e.g. the TCP endpoint of the media processor of
the bot or a new notification URL for further notifications regarding the call. The recording bot
application receives notifications via the aforementioned URL for users joining, leaving, muting,
starting screen shares and more.

After the bot recording application answered and accepted a call, the Microsoft Teams platform
opens a connection on the provided TCP endpoint. After an initial handshake for authorization
and encryption, metadata and media events are transferred via the TCP connection.

> [!IMPORTANT]  
> The TCP endpoint also requires a certificate signed by a valid authority. The Certificate for the
> HTTPS endpoint and the TCP endpoint **can** be the same.

## Graph Communications SDK

The overhead of the TCP endpoints is completly managed by the Graph Communications SDK, but the
HTTPS endpoints for notifications must be custom implemented by the bot recording application.
ASP.NET Core can be used. After a HTTP request is authorized and notifications are parsed from the

> request, the notifications should be forwarded to the SDK. The SDK will process notification and
> fire eventss based on these notifications. The event handlers can implement business logic. An

Reviewer 2: something went wrong there
```suggestion
request, the notifications should be forwarded to the SDK. The SDK will process notifications and
fire events based on these notifications. The event handlers can implement business logic. An
```
